src/instructlab/training/logger.py

```diff
@@ -0,0 +1,177 @@
+
+try:
+    # Third Party
+    from torch.utils.tensorboard import SummaryWriter
+except ImportError:
+    SummaryWriter = None
+
+# First Party
+from instructlab.training import async_logger
+
+
+def flatten_dict(
+    d: Mapping,
+    sep: str = ".",
+    prefix: str = "",
+    _flattened: dict | None = None,
+):
+    if _flattened is None:
+        _flattened = {}
+
+    for k, v in d.items():
+        if isinstance(v, Mapping):
+            flatten_dict(v, sep=sep, prefix=f"{prefix}{k}{sep}", _flattened=_flattened)
+        else:
+            _flattened[prefix + k] = v
+
+    return _flattened
+
+
+class BaseLogger:
+    def __init__(self, run_name: str | None, log_dir: str | Path):
+        if run_name is None:
+            # default template
+            run_name = "{time}_rank{rank}"
+
+        substitutions = {
+            "{time}": datetime.now().isoformat(),
+            "{rank}": os.environ.get("RANK", 0),
+            "{local_rank}": os.environ.get("LOCAL_RANK", 0),
+        }
+        for placeholder_pat, value in substitutions.items():
+            run_name = run_name.replace(placeholder_pat, str(value))
+
+        self.run_name = run_name
+        self.log_dir = log_dir
+
+    def setup(self):
+        pass
+
+    def log_dict(self, d: dict, step: int | None = None):
+        pass
+
+    def teardown(self, exit_code: int = 0):
+        pass
+
+
+class FileLogger(BaseLogger):
+    def setup(self):
+        filename = Path(self.log_dir) / f"{self.run_name}.jsonl"
+        os.makedirs(filename.parent, exist_ok=True)
+
+        self.file = open(filename, "a")  # pylint: disable=R1732
+
+    def log_dict(self, d: dict, step: int | None = None):
+        if step is not None and not "step" in d:
+            d["step"] = step
+
+        self.file.write(json.dumps(d, indent=None) + "\n")
+
+    def teardown(self, exit_code: int = 0):
+        self.file.close()
+
+
+class AsyncStructuredLogger(BaseLogger):
+    def __init__(self, run_name: str | None, log_dir: str | Path):
+        if run_name is None:
+            # Async custom template for backwards compat
+            run_name = "training_params_and_metrics_global{rank}"
+
+        super().__init__(run_name, log_dir)
+
+    def setup(self):
+        filename = Path(self.log_dir) / f"{self.run_name}.jsonl"
+        os.makedirs(filename.parent, exist_ok=True)
+        self.struct_logger = async_logger.AsyncStructuredLogger(filename)
+
+    def log_dict(self, d: dict, step: int | None = None):
+        self.struct_logger.log_sync(d)
+
+    def teardown(self, exit_code: int = 0):
+        pass
+
+
+class WandbLogger(BaseLogger):
+    def __init__(self, run_name: str | None, log_dir: str | Path):
+        super().__init__(run_name, log_dir)
+
+        if wandb is None:
+            msg = (
+                "Could not initialize WandbLogger because package wandb could not be imported. \
+                \nPlease ensure it is installed by running 'pip install wandb'"
+            )
```

Comment:
To make this more readable, you can make this multiline instead of escaping with `\` 

```suggestion
            msg = (
                "Could not initialize WandbLogger because package wandb could not be imported.\n"
                "Please ensure it is installed by running 'pip install wandb'"
            )
```